server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,464 @@
+import static com.google.common.collect.FluentIterable.from;
+import static java.lang.Math.max;
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static org.sonar.server.computation.component.ComponentVisitor.Order.POST_ORDER;
+
+public class FileMoveDetectionStep implements ComputationStep {
+  private static final Logger LOG = Loggers.get(FileMoveDetectionStep.class);
+  private static final int MIN_REQUIRED_SCORE = 90;
+  private static final List<String> FILE_QUALIFIERS = asList(Qualifiers.FILE, Qualifiers.UNIT_TEST_FILE);
+  private static final List<String> SORT_FIELDS = singletonList("name");
+
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private final TreeRootHolder rootHolder;
+  private final DbClient dbClient;
+  private final SourceLinesRepository sourceLinesRepository;
+  private final FileSimilarity fileSimilarity;
+  private final MutableMovedFilesRepository movedFilesRepository;
+
+  public FileMoveDetectionStep(AnalysisMetadataHolder analysisMetadataHolder, TreeRootHolder rootHolder, DbClient dbClient,
+    SourceLinesRepository sourceLinesRepository, FileSimilarity fileSimilarity, MutableMovedFilesRepository movedFilesRepository) {
+    this.analysisMetadataHolder = analysisMetadataHolder;
+    this.rootHolder = rootHolder;
+    this.dbClient = dbClient;
+    this.sourceLinesRepository = sourceLinesRepository;
+    this.fileSimilarity = fileSimilarity;
+    this.movedFilesRepository = movedFilesRepository;
+  }
+
+  @Override
+  public String getDescription() {
+    return "Detect file moves";
+  }
+
+  @Override
+  public void execute() {
+    // do nothing if no files in db (first analysis)
+    Snapshot baseProjectSnapshot = analysisMetadataHolder.getBaseProjectSnapshot();
+    if (baseProjectSnapshot == null) {
+      LOG.debug("First analysis. Do nothing.");
+      return;
+    }
+
+    Map<String, ComponentDtoWithSnapshotId> dbFilesByKey = getDbFilesByKey(baseProjectSnapshot);
+
+    if (dbFilesByKey.isEmpty()) {
+      LOG.debug("Previous snapshot has no file. Do nothing.");
+      return;
+    }
+
+    Map<String, Component> reportFilesByKey = getReportFilesByKey(this.rootHolder.getRoot());
+
+    if (reportFilesByKey.isEmpty()) {
+      LOG.debug("No files in report. Do nothing.");
+      return;
+    }
+
+    Set<String> addedFileKeys = ImmutableSet.copyOf(Sets.difference(reportFilesByKey.keySet(), dbFilesByKey.keySet()));
+    Set<String> removedFileKeys = ImmutableSet.copyOf(Sets.difference(dbFilesByKey.keySet(), reportFilesByKey.keySet()));
+
+    // can find matches if at least one of the added or removed files groups is empty => abort
+    if (addedFileKeys.isEmpty() || removedFileKeys.isEmpty()) {
+      LOG.debug("Either no files added or no files removed. Do nothing.");
+      return;
+    }
+
+    // retrieve file data from db and report
+    Map<String, File> dbFileSourcesByKey = getDbFileSourcesByKey(dbFilesByKey, removedFileKeys);
```

Comment:
ok to discuss about that later. Note that this is the strategy decided and applied in scanners (FS is used, through Persistit cache or protobuf files, when size of computation data depends on project).